aws/rust-runtime/aws-config/src/sts/assume_role.rs

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Assume credentials for a role through the AWS Security Token Service (STS).
+
+use aws_sdk_sts::error::AssumeRoleErrorKind;
+use aws_sdk_sts::{Client, Config};
+use aws_types::credentials::{self, future, CredentialsError, ProvideCredentials};
+use aws_types::region::Region;
+
+use tracing::Instrument;
+
+/// Credentials provider that uses credentials provided by another provider to assume a role
+/// through the AWS Security Token Service (STS).
+///
+/// When asked to provide credentials, this provider will first invoke the inner credentials
+/// provider to get AWS credentials for STS. Then, it will call STS to get assumed credentials for
+/// the desired role.
+///
+/// # Examples
+/// ```rust
+/// use aws_config::sts::{AssumeRoleProvider};
+/// use aws_types::{Credentials, region::Region};
+/// use aws_config::environment;
+/// use aws_config::environment::credentials::EnvironmentVariableCredentialsProvider;
+///
+/// let provider = AssumeRoleProvider::new(
+///     EnvironmentVariableCredentialsProvider::new(),
+///     "arn:aws:iam::123456789012:role/demo",
+///     Region::from_static("us-east-2"),
+///     "testAR",
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct AssumeRoleProvider<P> {
```

Comment:
for consistency with all the other providers, we should at least register a TODO to create a builder and create a constructor that has a method that accepts a `ProviderConfig`

Reply:
I didn't provide a builder currently given that all of the fields are required. Builders aren't great at expressing many required parameters (like here), since it's hard to actually denote that they're required at compile-time. It's possible through generics tricks, but it's not very nice. We may want a builder anyway in case we add support for more fields later (like `ExternalId`), but I think the builder's constructor should still take all of these arguments.

Reply:
these parameters are sort of required (but also sort of aren't):
- Region can be loaded from the default chain, this is what most customers will use instead of manually specifying a region
- We will need the ability to override the interior HTTP connection, both for supporting non-hyper implementations & also for integration testing where we use an in memory connection

Reply:
Ah, I was running into issues with the default chain giving me `NoProvider` errors when I didn't set it explicitly, but I'm guessing that's because I was setting my AWS credentials through envvars, not the config, which doesn't also set the region 👍 

Also added the ability to override the connection.